src/main/scala/is/hail/expr/types/TVariable.scala

```diff
@@ -15,24 +18,20 @@ final case class TVariable(name: String, var t: Type = null) extends Type {
     throw new RuntimeException("TVariable is not realizable")
 
   override def unify(concrete: Type): Boolean = {
-    if (t == null) {
-      if (concrete.isRealizable) {
-        t = concrete
-        true
-      } else
-        false
-    } else
-      t.isOfType(concrete)
+    if ((b.isDefined || concrete.isRealizable) && cond(concrete))
```

Comment:
I don't think you need isDefined here.  The unify of box now takes care of that.

Reply:
In particular I think you can just write concrete.isRealizable() && cond(concrete) && b.unify(concrete)

Reply:
oh, nice